runners/core-java/src/main/java/org/apache/beam/runners/core/InMemoryMultimapSideInputView.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+  */
+package org.apache.beam.runners.core;
+
+import com.google.common.base.Objects;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multimaps;
+import java.util.Collections;
+import org.apache.beam.sdk.coders.Coder;
+import org.apache.beam.sdk.transforms.Materializations;
+import org.apache.beam.sdk.transforms.Materializations.MultimapView;
+import org.apache.beam.sdk.values.KV;
+
+/**
+ * An in-memory representation of {@link MultimapView}.
+ */
+public class InMemoryMultimapSideInputView<K, V> implements Materializations.MultimapView<K, V> {
+
+  /**
+   * Creates a {@link MultimapView} from the provided values. The provided {@link Coder} is used
+   * to guarantee structural equality for keys instead of assuming Java object equality.
+   */
+  public static <K, V> MultimapView<K, V> fromIterable(
+      Coder<K> keyCoder, Iterable<KV<K, V>> values) {
+    // We specifically use an array list multimap to allow for:
+    //  * null keys
+    //  * null values
+    //  * duplicate values
+    Multimap<Object, Object> multimap = ArrayListMultimap.create();
```

Comment:
This can be a `ListMultimap` to force a bit more of the preconditions. Up to you.

Reply:
The List part isn't important. Its just the ability to store null keys and values and duplicate values. For example HashMultimap doesn't support duplicate values and ImmutableMultimap doesn't support null keys/values.

Reply:
I wanted to use `ListMultimap` instead of generic `Multimap` to signal the non-set nature of the `Multimap`, as there's unfortunately no `BagMultimap`, but It's a minor nit given the commenting.